lib/logstash/outputs/s3.rb

```diff
@@ -397,16 +398,48 @@ def restore_from_crash
     @crash_uploader = Uploader.new(bucket_resource, @logger, CRASH_RECOVERY_THREADPOOL)
 
     temp_folder_path = Pathname.new(@temporary_directory)
-    Dir.glob(::File.join(@temporary_directory, "**/*"))
-      .select { |file| ::File.file?(file) }
-      .each do |file|
-      temp_file = TemporaryFile.create_from_existing_file(file, temp_folder_path)
-      if temp_file.size > 0
-        @logger.debug? && @logger.debug("Recovering from crash and uploading", :path => temp_file.path)
-        @crash_uploader.upload_async(temp_file, :on_complete => method(:clean_temporary_file), :upload_options => upload_options)
+
+    files = Dir.glob(::File.join(@temporary_directory, "**/*"))
+               .select { |file_path| ::File.file?(file_path) }
+    under_recovery_files = get_under_recovery_files(files)
+
+    files.each do |file_path|
+      # when encoding is GZIP, if file is already recovering or recovered and uploading to S3, log and skip
+      if under_recovery_files.include?(file_path)
+        unless file_path.include?(TemporaryFile.gzip_extension)
+          @logger.warn("The #{file_path} file either under recover process or failed to recover before.")
+        end
       else
-        clean_temporary_file(temp_file)
+        temp_file = TemporaryFile.create_from_existing_file(file_path, temp_folder_path)
+        if temp_file.size > 0
+          @logger.debug? && @logger.debug("Recovering from crash and uploading", :path => temp_file.path)
+          @crash_uploader.upload_async(temp_file,
+                                       :on_complete => method(:clean_temporary_file),
+                                       :upload_options => upload_options)
+        end
+        # do not remove if Logstash tries to recover but fails
+        if @encoding != GZIP_ENCODING && temp_file.size != 0
```

Comment:
I think it might be worthwhile to see some tests here - what happens with:

- standard plain text recovery
- standard gzip recovery
- empty plain text recovery
- empty gzip recovery

Reply:
It seems default (`none` or `plain txt`) encoding cases are covered in the integration test, [`restore_from_crash_spec.rb`](https://github.com/logstash-plugins/logstash-output-s3/blob/d96531b573a8ff4e2cc216ad1c8cb83027000499/spec/integration/restore_from_crash_spec.rb) and I have added the cases where Logstash restores the file, uploads and removes locally.

- Run integration test, as we synced offline, make sure to set the AWS ENV params including `AWS_SESSION_TOKEN`:
`bundle exec rspec logstash-plugins/logstash-output-s3/spec/integration/restore_from_crash_spec.rb:69 -t integration:true`
- Result logs:
```
Finished in 1 minute 22.39 seconds (files took 9.97 seconds to load)
3 examples, 0 failures
```